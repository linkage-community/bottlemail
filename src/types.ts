type Node<T> = {
  kind: T,
  value: string,
  raw: string,
}

export const TextKind = 'Text'
// raw, value are same
export type Text = Node<typeof TextKind>

export const EmojiNameKind = 'EmojiName'
// eg: raw => ':smile:', value => 'smile'
export type EmojiName = Node<typeof EmojiNameKind>

export const LinkKind = 'Link'
// raw, value are same
export type Link = Node<typeof LinkKind>

export const MentionKind = 'Mention'
// eg: raw => '@dolphin', value => 'dolphin'
export type Mention = Node<typeof MentionKind>

export type NodeType = Link | Text | Mention | EmojiName

// helpers
export const isText = (n: NodeType): n is Text => {
  return n.kind === TextKind
}
export const isEmojiName = (n: NodeType): n is EmojiName => {
  return n.kind === EmojiNameKind
}
export const isLink = (n: NodeType): n is Link => {
  return n.kind === LinkKind
}
export const isMention = (n: NodeType): n is Mention => {
  return n.kind === MentionKind
}
