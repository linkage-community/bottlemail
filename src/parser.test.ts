import { MentionKind, TextKind, EmojiNameKind } from "./types"
import parse from "./parser"

import pictograph from "pictograph"

describe("Parser", () => {
	it("must return `[]` when no argument given.", () => {
		// JavaScript user can use
		expect((parse as any)()).toEqual([])
	})

	it("must parse empty string as `[]`.", () => {
		expect(parse("")).toEqual([])
	})

	describe("complex cases", () => {
		const testcase1 =
			"@otofune Yo! :smile: https://github.com/ http://[fe80::a1b3:125d:c1f8:4780]/ @ @test"
		test(`must parse '${testcase1}' as equivalent to the snapshot.`, () => {
			expect(parse(testcase1)).toMatchSnapshot()
		})

		const testcase2 =
			"This includes special identifiers :aa @*  but all are invalid as Emoji, Mention."
		test(`must parse '${testcase2}' as '${TextKind}' kind node.`, () => {
			expect(parse(testcase2)).toEqual([
				{ kind: TextKind, value: testcase2, raw: testcase2 }
			])
		})
	})

	describe("simple cases", () => {
		const testcase1 = "@dolphin"
		const testcase1Expected = "dolphin"
		it(`must parse '${testcase1}' as node with value '${testcase1Expected}'`, () => {
			expect(parse(testcase1)).toEqual([
				{ kind: MentionKind, value: testcase1Expected, raw: testcase1 }
			])
		})

		describe("about " + EmojiNameKind, () => {
			for (const name of Object.keys(pictograph.dic)) {
				const emojiName = `:${name}:`
				it(`must parse '${emojiName}' as ${EmojiNameKind} node with value '${name}'`, () => {
					expect(parse(emojiName)).toEqual([
						{
							kind: EmojiNameKind,
							value: name,
							raw: emojiName
						}
					])
				})
			}
		})
	})

	it(`"(@dolphin hi)" correctly parse`, () => {
		expect(parse("(@dolphin hi)")).toEqual([
			{ kind: TextKind, value: "(", raw: "(" },
			{ kind: MentionKind, value: "dolphin", raw: "@dolphin" },
			{ kind: TextKind, value: " hi)", raw: " hi)" }
		])
	})

	const nonURLStartsWithHTTP = "httpisgod"
	it(`parse "${nonURLStartsWithHTTP}" as Text`, () => {
		expect(parse(nonURLStartsWithHTTP)).toEqual([
			{ kind: TextKind, value: "httpisgod", raw: "httpisgod" }
		])
	})
})
