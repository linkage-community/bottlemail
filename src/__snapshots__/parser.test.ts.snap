// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser must parse '@otofune Yo! :smile: https://github.com/ http://[fe80::a1b3:125d:c1f8:4780]/ @ @test' as equivalent to the snapshot. 1`] = `
Array [
  Object {
    "kind": "Mention",
    "raw": "@otofune",
    "value": "otofune",
  },
  Object {
    "kind": "Text",
    "raw": " Yo! ",
    "value": " Yo! ",
  },
  Object {
    "kind": "EmojiName",
    "raw": ":smile:",
    "value": "smile",
  },
  Object {
    "kind": "Text",
    "raw": " ",
    "value": " ",
  },
  Object {
    "kind": "Link",
    "raw": "https://github.com/",
    "value": "https://github.com/",
  },
  Object {
    "kind": "Text",
    "raw": " ",
    "value": " ",
  },
  Object {
    "kind": "Link",
    "raw": "http://[fe80::a1b3:125d:c1f8:4780]/",
    "value": "http://[fe80::a1b3:125d:c1f8:4780]/",
  },
  Object {
    "kind": "Text",
    "raw": " @ ",
    "value": " @ ",
  },
  Object {
    "kind": "Mention",
    "raw": "@test",
    "value": "test",
  },
]
`;
